File: src/features/presence/domain/entities/presence.ts
/** The four presence types, in display order. */
export const PRESENCE_TYPES = [
  "office",
  "remote",
  "vacation",
  "holiday",
] as const;

export type PresenceType = (typeof PRESENCE_TYPES)[number];

export interface Presence {
  id: string;
  profileId: string;
  /** Epoch ms at UTC midnight of the day concerned. */
  day: number;
  type: PresenceType;
  /** Day commute footprint in kgCO2e. `null` when not computed (imported days,
   * or days never encoded with the CO2 feature); `0` is a real computed value
   * (e.g. télétravail with no trip). */
  co2Kg: number | null;
  /** True if any of the day's trips fell back to an estimated factor. */
  isEstimated: boolean;
  createdAt: number;
  updatedAt: number;
  /** Sum of minutes of work entries for this day. */
  workMinutes: number;
}

/** One commute leg attached to a presence day. Structurally identical to the
 * commute feature's `TripInput`; declared here so presence stays decoupled. */
export interface PresenceTrip {
  modeId: string;
  distanceKm: number;
  roundTrip: boolean;
  occupants: number;
}

/** Fields supplied when setting a presence (`id`/timestamps are backend-managed).
 * `trips` carries the commute legs for office/remote days (cleared otherwise). */
export interface SetPresenceInput {
  profileId: string;
  day: number;
  type: PresenceType;
  trips?: PresenceTrip[];
}

File: src-tauri/src/domain/entities/presence.rs
use crate::domain::error::DomainError;

/// The kind of a day's presence. Stored as a lowercase string in SQL; the
/// domain owns the (de)serialization and the validation of unknown values.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PresenceType {
    Office,
    Remote,
    Vacation,
    Holiday,
}

impl PresenceType {
    /// Lowercase wire/storage representation.
    pub fn as_str(&self) -> &'static str {
        match self {
            PresenceType::Office => "office",
            PresenceType::Remote => "remote",
            PresenceType::Vacation => "vacation",
            PresenceType::Holiday => "holiday",
        }
    }

    /// Parse a stored/incoming string into a `PresenceType`, rejecting unknowns.
    pub fn parse(value: &str) -> Result<Self, DomainError> {
        match value {
            "office" => Ok(PresenceType::Office),
            "remote" => Ok(PresenceType::Remote),
            "vacation" => Ok(PresenceType::Vacation),
            "holiday" => Ok(PresenceType::Holiday),
            other => Err(DomainError::Validation(format!(
                "invalid presence type: {other}"
            ))),
        }
    }
}

/// A single day of presence for a profile (lives in the encrypted vault).
#[derive(Debug, Clone)]
pub struct Presence {
    pub id: String,
    pub profile_id: String,
    /// Epoch ms at UTC midnight of the day concerned.
    pub day: i64,
    pub kind: PresenceType,
    /// Denormalized commute footprint for the day (kgCO2e). `None` when the day
    /// carries no trip (e.g. vacation/holiday, or a pre-CO2 imported row).
    pub co2_kg: Option<f64>,
    /// `true` if any of the day's trips fell back to an estimated factor (R6).
    pub is_estimated: bool,
    pub created_at: i64,
    pub updated_at: i64,
}
